plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSegmentPruner.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spooling.filesystem;
+
+import com.google.inject.Inject;
+import io.airlift.log.Logger;
+import io.airlift.units.Duration;
+import io.trino.filesystem.FileEntry;
+import io.trino.filesystem.FileIterator;
+import io.trino.filesystem.Location;
+import io.trino.filesystem.TrinoFileSystem;
+import io.trino.filesystem.TrinoFileSystemFactory;
+import io.trino.spi.security.ConnectorIdentity;
+import jakarta.annotation.PostConstruct;
+import jakarta.annotation.PreDestroy;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ScheduledExecutorService;
+
+import static io.trino.spooling.filesystem.FileSystemSpooledSegmentHandle.fromStorageObjectName;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class FileSystemSegmentPruner
+{
+    private final Logger log = Logger.get(FileSystemSegmentPruner.class);
+
+    private final TrinoFileSystem fileSystem;
```

Comment:
```
All major cloud storage provides (GCS, S3, ABFS) list files in lexicographical order
which makes it possible to iterate over files until the first non-expired segment is reached.
 ```

Is this part of public interace or just an observation. Also I am not sure we should assume that this is part of `TrinoFileSystem` interface. If so we should explicitly expose that files are listed in alphabetical order as some property of given TrinoFileSystem implementation and check the flag here. If it is not present we can list everything and sort locally.

Reply:
It's a part of the public interface.

Reply:
S3 (https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html):

<img width="858" alt="Screenshot 2024-09-09 at 09 52 33" src="https://github.com/user-attachments/assets/e15c82b1-7d21-4794-92e1-5a92d3ac0946">

GCS (https://cloud.google.com/storage/docs/listing-objects):

<img width="803" alt="Screenshot 2024-09-09 at 09 53 19" src="https://github.com/user-attachments/assets/b78b4b80-165f-4d80-9f85-aa8c431ea6b1">

ABFS (https://learn.microsoft.com/en-us/rest/api/storageservices/list-blobs?tabs=microsoft-entra-id):

<img width="682" alt="Screenshot 2024-09-09 at 09 54 30" src="https://github.com/user-attachments/assets/1ec4243d-cc1a-4574-ad20-9774068ec7aa">

Reply:
The generated ULIDs are upper-case only and lexographically ordered by the timestamp which makes it usable for this particular use-case on all cloud storages.

Reply:
> Also I am not sure we should assume that this is part of TrinoFileSystem interface. If so we should explicitly expose that files are listed in alphabetical order as some property of given TrinoFileSystem implementation and check the flag here. If it is not present we can list everything and sort locally.

And what about this one? :)

Reply:
I'm aware but it seems like a stretch in a context of this PR. We only support S3, GCS, Azure which has this semantics. I can improve filesystem documentation/tests as a follow up

Reply:
I've added a logic that detects a broken order and switches to "scan all files" logic